### ISSUE TITLE
Sidebar permalink panel: consider to always render the panel

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#12031]
Splitting this out from #11874.

When creating a new post, the new permalink panel introduced in #11874 is not rendered yet in the sidebar.

![screenshot 2018-11-18 at 14 58 32](https://user-images.githubusercontent.com/1682452/48673546-b1e8e800-eb42-11e8-904b-c6772ce8efbf.png)

The panel appears only after the first auto-save or after users click on "Save draft" or "Publish".

![screenshot 2018-11-18 at 14 59 11](https://user-images.githubusercontent.com/1682452/48673549-b6ad9c00-eb42-11e8-99f3-3b711014b001.png)

This could be confusing for users who are just exploring the new post screen and wondering where the permalink setting is. It could be even more confusing for screen reader users, as screen readers have special tools to explore a web page (e.g. list all the headings, list all the form controls, list all the links, etc.) and they won't find anything related to the permalink setting. Same potential confusion applies to low vision users, screen magnifier users, users with cognitive impairments, etc.

I'd suggest to consider to always show the panel in the sidebar with some meaningful message when a slug doesn't exist yet.

**Comment by afercia** [comment]
Also, if I remember correctly, it was decided to remove the permalink UI in the post title. Not sure it's already tracked down in some other issue. /Cc @youknowriad

**Comment by youknowriad** [comment]
Removing the permalink UI in the post title had my preference but I think there were some hesitation. I'd love thoughts from others @jasmussen @karmatosed @mapk @chrisvanpatten @earnjam

**Comment by afercia** [comment]
I'd vote for removing it. It has serious accessibility issues. I'd say also usability and discoverability aren't ideal.

**Comment by afercia** [comment]
Worth reminding this issue with more details about the a11y issues: https://github.com/WordPress/gutenberg/issues/5494

It was closed only because there's now the new panel. However, the old UI is still not accessible.

> We didn't remove this UI yet because we're passed UI freeze but it should be considered deprecated.

**Comment by jasmussen** [comment]
Considering the flow that a casual user might want to dismiss the sidebar and never open it again, it would be nice if the permalink was editable in the pre-publish flow (i.e. press Publish... and the permalink could be a panel there). If we had that flow in place then it would be harder to find good arguments to keeping the current in-canvas interface in place.

**Comment by earnjam** [comment]
#12009 has ability to edit at any time and would resolve this issue. It has permalink sidebar panel visible but not expanded on initial load. 

If that lands it would also make adding a permalink editor to the pre-publish panel pretty easy. It would be hard to do without it because if a user publishes without first saving a draft, we don’t know the permalink structure. 

Main consideration is it requires 2 core changes. I’ll work on it today and get patch up for the 2nd core ticket. I also need to make a small update to the PR so it can work correctly in the plugin until the core patch lands.

**Comment by mapk** [comment]
I believe our publishing flow causes a bit of confusion here. I've seen more than one occassion where the user clicks the Publish button expecting to publish the post regardless of the ellipses at the end of the button or not. It's surprising to them that a pre-publish sidebar appears, and I've seen them just quickly click "Publish" button again thinking something went not as expected.

That being said, having the Permalink section in the Doc Inspector makes more sense to me. This way a user wouldn't have to commit to "Publishing" their page before seeing another sidebar that allows them to make that adjustment. 

My vote is to keep it in the Doc Inspector, and to show it from the beginning as suggested by @afercia in this issue.

**Comment by afercia** [comment]
> I believe our publishing flow causes a bit of confusion here 

Yep, also from an accessibility perspective the behavior of the Publish button / panel is a concern. The first implementation was discussed at length in #4187 and improved a bit. Still, the way the three steps work can be highly confusing for assistive technologies users: something appears on the page, focus is moved there unexpectedly, there's nothing to inform there are publishing options _before_ getting there, etc.

Also, worth nothing the pre-publish checks can be disabled. Placing important settings _only_ there is not advisable because if users disable this panel, some settings could be completely unavailable:

<img width="284" alt="screenshot 2019-02-28 at 08 45 36" src="https://user-images.githubusercontent.com/1682452/53550351-c7ed4800-3b36-11e9-9a61-aabe1828d13e.png">

Out of the scope of this issue but maybe worth trying to revive the explorations on #7602 /Cc @mapk

**Comment by jasmussen** [comment]
> Placing important settings _only_ there is not advisable because if users disable this panel, some settings could be completely unavailable. 

For that reason it's also in Options, along with Tips:

![screenshot 2019-02-28 at 09 25 44](https://user-images.githubusercontent.com/1204802/53552078-dd647100-3b3a-11e9-9309-f17cc74ebf57.png)

**Comment by afercia** [comment]
@jasmussen I know, but that's not the point 🙂 Was just advising to not put the permalink UI only in the pre-publish checks. As I see it, the pre-publish checks should contain only optional _suggestions_ and not important features like permalinks.

**Comment by jasmussen** [comment]
Of course they should be redundant UI, I hope no-one has suggested otherwise.

**Comment by earnjam** [comment]
@afercia, I believe @jasmussen was just suggesting adding a way to edit permalinks to the pre-publish panel to support removing it from the title area. The sidebar panel would remain. Similar to how visibility and date are done now.

**Comment by jasmussen** [comment]
Yes, absolutely. To clarify because there was confusion (sorry about that):

1. Always in sidebar: cool
2. Remove from editing canvas: cool if we add it also to the pre-publish dialog

If we remove it from the editing canvas and have it _only_ in the sidebar, then users who never use the sidebar won't have access to it. By putting an extra copy of the permalink UI in the pre-publish check, it can be a "last minute" way to tweak those final things before publishing.

**Comment by afercia** [comment]
Makes sense to me 🙂 
Maybe also consider to add a button "change permalink" to the _post-publish_ check?

<img width="279" alt="screenshot 2019-02-28 at 13 48 17" src="https://user-images.githubusercontent.com/1682452/53567515-ba988380-3b5f-11e9-9adf-72318e90486c.png">

That would be the last chance, as in: "Is the permalink OK? want to change it"?

Also, I'd tend to think the readonly field above should display the _last_ part of the "address".

**Comment by jasmussen** [comment]
In the post-publish check it's already "too late" in many cases to change it. For example a plugin might have already shared the post on social media. What would be the problem to solve? Similarly, the post address is a grayed field because the intent is for you to be able to copy that link directly and share it wherever you like, because it's now _live_. The UI isn't necessarily perfect for it, but that's the history of how we got there.

**Comment by afercia** [comment]
> it's already "too late" ... For example a plugin might have already shared the post on social media

There are redirects to solve this issue. Also, the permalink can be edited at any time after the post has been published, and the permalink structure itself can be changed at any time. WordPress takes care of this.

> What would be the problem to solve?

SEO for example. As an user, I might realize my permalink is not ideal only after I published the post and want to change it.

> Similarly, the post address is a grayed ...

Not arguing if it's grayed / readonly.

I argue it displays the first part of the URL, which doesn't help me at all.

<img width="279" alt="screenshot 2019-02-28 at 13 48 17" src="https://user-images.githubusercontent.com/1682452/53569092-a6ef1c00-3b63-11e9-9de3-6c828f2dd853.png">

I don't need to see my address starts with (in this case) `http://build.wordpress-develop.test/`. That's my site domain, I know it already 🙂 Instead, I need to see the last part of the URL.

**Comment by jasmussen** [comment]
Although that can work, it still feels slightly confusing to suggest the permalink can be edited after the fact, and if it's that important for SEO then seems worth giving more presence to the panel in the pre publish state.

Regardless, both those screens should be heavily pluggable so plugins could do this.

Could we make the link field wrap instead of crop the URL? Seems like it'd be misleading if the copy button copied the full URL but only a part of it was shown.

All of the above is separate to this ticket though.

**Comment by amber-c** [comment]
Chiming in to say that I supremely dislike the inability to see/edit the slug prior to saving a post or page with a passion. 

As a more advanced user I find it so counter productive to name a page/post, know that I want to customize that slug to something specific, and have no way to do it other than publishing (i.e Publish/Pending) and then go back and edit it. It's just extra unnecessary disruptive steps.

**Comment by mapk** [comment]
I'm good with showing the Permalink module in the Document Inspector at all times and from the very beginning of creating the draft. Maybe we just set default to collapsed.

**Comment by earnjam** [comment]
With #12009 the permalink panel now shows on initial editor load, so I think this is resolved. 

@afercia feel free to reopen if there is still something to address here.

**Comment by afercia** [comment]
Regarding the old permalink UI (the one _above the title_), worth reminding #5494 is still open and waiting to be solved since March 2018, which is two years now.

The old UI has inherent accessibility problems, starting with its placement before the title. At some point [it was declared deprecated](https://github.com/WordPress/gutenberg/issues/5494#issuecomment-439104668). Long time has passed since then and we were only waiting for a new UI to be complete and fully functional. We should now either remove the old UI or change it drastically to solve all its issues.
